python_modules/dagster/dagster/core/storage/runs/sql_run_storage.py

```diff
@@ -330,9 +343,70 @@ def get_runs(
         limit: int = None,
         bucket_by: Optional[Union[JobBucket, TagBucket]] = None,
     ) -> List[PipelineRun]:
-        query = self._runs_query(filters, cursor, limit, bucket_by=bucket_by)
-        rows = self.fetchall(query)
-        return self._rows_to_runs(rows)
+        return [
+            self._row_to_run(row)
+            for row in self._get_rows(
+                filters=filters, limit=limit, cursor=cursor, bucket_by=bucket_by
+            )
+        ]
+
+    def _get_rows(
+        self,
+        filters=None,
+        limit=None,
+        columns=None,
+        order_by=None,
+        ascending=False,
+        cursor=None,
+        bucket_by=None,
+    ):
+        if not bucket_by or self.supports_bucket_queries:
+            query = self._runs_query(
+                filters=filters,
+                limit=limit,
+                columns=columns,
+                order_by=order_by,
+                ascending=ascending,
+                cursor=cursor,
+                bucket_by=bucket_by,
```

Comment:
this is sort of tricksy indirection here - we pass the bucket_by through if its supported other wise we work around here in this function just called `_get_rows`. Some ideas to make this clearer

* rename this fn to something a bit more descriptive
* assert in _runs_query when bucket_by is set but support_bucket is not - refer back to this fn in the error message 
* explanatory comments

Reply:
Yeah, I have another version where all of this checking/gating logic takes place in the loader.  It's much much cleaner, but it's a bit odd to have the support logic live in the storage layer, but the bucket calls happening outside of storage.